`asarray`: take the default device into consideration.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106779

Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

ysiraichi committed:
`asarray`: take the default device into consideration.

Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "`asarray`: take the default device into consideration."


Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

ysiraichi committed:
Update on "`asarray`: take the default device into consideration."


Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

ysiraichi committed:
Revert change in implementation. on "`asarray`: take the default device into consideration."


Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

ysiraichi committed:
Add tests. on "`asarray`: take the default device into consideration."


Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

rgommers reviewed (approved):
LGTM, thanks for the quick fix @ysiraichi!

ysiraichi commented on test/test_tensor_creation_ops.py:
```diff
@@ -3966,6 +3966,29 @@ def test_numpy_scalars(self, device):
+            3,
+            np.arange(5),
+            torch.arange(5),
+            array.array("f", [1, 2, 3, 4]),
```
Ah, you are right. I used `"f"` here because:

ysiraichi committed:
Make test more robust. on "`asarray`: take the default device into consideration."


Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge

ysiraichi committed:
Rebased. on "`asarray`: take the default device into consideration."


Fix: #106773

This PR makes it so `asarray` takes the default device into consideration when called with
a Python sequence as the data.

[ghstack-poisoned]

ysiraichi commented:
CI failures are not related.
